anvil/config.py

```diff
@@ -105,20 +115,9 @@ def produce_checkpoint(self, cp_id=None, training_output=None):
 
     def produce_training_context(self, training_output):
         """Given a training output produce the context of that training"""
-        with self.set_context(ns=self._curr_ns.new_child(training_output.as_input())):
-            _, geometry = self.parse_from_(None, "geometry", write=False)
-            _, model = self.parse_from_(None, "model", write=False)
-            _, target = self.parse_from_(None, "target", write=False)
-            _, cps = self.parse_from_(None, "checkpoints", write=False)
-            _, base = self.parse_from_(None, "base", write=False)
-
-        return dict(
-            geometry=geometry,
-            model=model,
-            target=target,
-            checkpoints=cps,
-            base=base,
-        )
+        # NOTE: This seems a bit hacky, exposing the entire training configuration
+        # file - hopefully doesn't cause any issues..
+        return training_output.as_input()
```

Comment:
What does this do exactly?

Reply:
as_input() just returns a parsed config file of the fit (which gets copied at start of fit). the function is called as_input because this means we can use "from_" in the runcard.

So basically here instead of manually filling in the dictionary as we were before we just return the entire dictionary used to define the training settings. I think it will be ok because I don't see us having a key which means one thing in the training and a different thing in the sampling - but we will need to be careful for that (key words like batch_size springs to mind..) if we define a parse function for every keyword we use in the training and the sampling then with any luck we should spot any problems like that

Now because we return the fit config like lattice_length, lattice_dimension etc. etc. then when we use the collect function the namespace has all of the necessary dependencies to produce things like the model, the target etc. we were doing that before in the set_context block so in that sense I think this is actually nicer.

Reply:
does that answer your question?

Reply:
> Sure - I understand "parsed config file" to mean big dictionary with the stuff parsed / produced using the training / "fitting" runcard. Just couldn't find this in the reportengine doc.
> 
> So at the moment we are making this dictionary available during sampling using `training_context`. Would an alternative be to look at every parameter / action we need during sampling, and write `action: from_ training_output` in the sample runcard? I realise 

Not sure about action but anything which is in phi4_train.yaml like `use_arxiv_version` or whatever

`use_arxiv_version: { from_: training output }`

> this is not preferable, and we couldn't use collect, but I'm just trying to understand it all better.

well this goes down the line of having a very verbose sample config file where everything says from_: training_output but isn't necessarily bad. I feel like we always have ran a training and then want to do an analysis where we don't have to retype all the training config and so returning the full training config as a namespace via this production rule seems sensible with the caveat that we have to be careful to not double define variables

> 
> Also, could we not get rid of `produce_training_geometry` if everything is now contained in `training_context`?

possibly, I haven't really thought about it. I will check before merge